sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InsertMapSortExpression.scala

```diff
@@ -77,15 +71,56 @@ object InsertMapSortInGroupingExpressions extends Rule[LogicalPlan] {
             }.asInstanceOf[NamedExpression]
         }
         val newChild = Project(child.output ++ exprToMapSort.values, child)
-        val newAgg = Aggregate(newGroupingKeys, newAggregateExprs, newChild)
+        val newAgg = Aggregate(newGroupingKeys, newAggregateExprs, newChild, hint)
         newAgg -> agg.output.zip(newAgg.output)
     }
   }
+}
+
+/**
+ * Adds [[MapSort]] to [[RepartitionByExpression]] expressions containing map columns,
+ * as the key/value pairs need to be in the correct order before repartitioning:
+ *
+ * SELECT * FROM TABLE DISTRIBUTE BY map_column =>
+ * SELECT * FROM TABLE DISTRIBUTE BY map_sort(map_column)
+ */
+object InsertMapSortInRepartitionExpressions extends Rule[LogicalPlan] {
```

Comment:
can we combine these two rules so that we only need to traverse the plan once?

Reply:
initially i wanted to do the same, but logic for `InsertMapSortInGroupingExpressions` and `InsertMapSortInRepartitionExpressions` are quite different: Grouping produced new output after applying the changes, while Repartition only updates existing `RepartitionByExpression` by replacing `partitionExpressions`. 
Also, there is a dependency between `InsertMapSortInGroupingExpressions` and `PullOutGroupingExpressions`, as mentioned in [this comment](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L320).
For those reasons i decided to split them into separate Rules. But if you think performance saving from reduced traverse will be significant, we can combine those sure.

Reply:
I think it's fine to use `transformUpWithNewOutput` for both. If we hit `RepartitionByExpression`, we return `Nil` as the new output.

Reply:
i think this will make things more complex, while there is no need to return new output for `RepartitionByExpression`. Also, to prevent traverse on every plan we have added two conditions in `InsertMapSortInRepartitionExpressions`:

1. `_.containsPattern(REPARTITION_OPERATION)` as `cond` to `transformUpWithPruning`.
2. ` if rep.partitionExpressions.exists(mapTypeExistsRecursively)` in case matching.

So i would keep these two independent from each other.